running-code/execute-package.md

```diff
@@ -0,0 +1,163 @@
+```
+:::
+
+### Further exploration
+
+On your own or in small groups:
+
+* Install the `my_program.py` module from the [last lesson](execute-script-launch-command)
+* Try to get the same greeting as before using `python -m my_program`.
+
+**my_program.py** which prints out `✨ Hello from Python ✨` when executed as a script.
+
+```python
+#!/usr/bin/env python
+# The above line is a shebang and can take the place of typing python on the command line
+# This comment is below, because shebangs must be the first line!
+
+def shiny_hello():
+    print("\N{Sparkles} Hello from Python \N{Sparkles}")
+
+if __name__ == "__main__":
+    shiny_hello()
+```
+
+### Executable packages
+
+The `-m` flag as described above only works for Python modules (files), but does not work for Python (sub-)packages (directories).
+This means that you cannot execute a command using only the name of your package when it is structured to use directories
+
+Once your package grows, the top-level name `my_program` turns into a directory.
+(See [Python Package Structure](https://www.pyopensci.org/python-package-guide/package-structure-code/python-package-structure.html)
+for when and how to create a package structure).
+```
+project/
+└── src/
+    └── my_program/
+        ├── __init__.py
+        └── greeting.py
+```
+
+However, you can never execute a directory.
+```bash
+python -m my_program
+python: No module named my_program.__main__; 'my_program' is a package and cannot be directly executed
+```
+
+Initially, Python seems to tell you that the directory names, including your top-level package name,
+cannot be directly executed. But the error message contains the hint that you need to make this run properly.
+
+[Earlier you learned](execute-script-name-eq-main) that `if __name__ == "__main__":` can protect parts of your
+script from executing when it is imported, making that conditional change the file's behavior when used as a script.
+There is a very similar concept that can be applied to Python packages.
+
+Any package that contains a [`__main__.py` module](https://docs.python.org/3/library/__main__.html#module-__main__)
+can be executed directly using the `python` command, without reference to any specific module files.
+
+:::{note}
+The `__main__.py` file typically doesn't have an `if __name__ == "__main__":` conditional in it, as its execution
+is already separated out from the rest of the package.
+:::
+
+### Further exploration
+
+- Try to separate `my_program` into a package containing two files, including one `__main__.py`
+- Try to get `python -m my_program` to work as before
+- Does importing `my_program` still work [as before the separation](execute-script-name-eq-main)?
+
+```python
+import my_program
+
+def guess_my_number():
+    my_program.shiny_hello()
+    print("Was your number 42?")
+
+guess_my_number()
+# ✨ Hello from Python ✨
+# Was your number 42?
+```
+
+:::{attention}
+Don't forget to (re)install your package after creating this file!
+:::
```

Comment:
hopefully we are encouraging editable installs when working on stuff? or maybe remind that here: "reinstall your package, unless you installed it as an editable install (link to wherever we introduce that)"

Reply:
We do in packaging tutorials. But this was originally designed as extending a Clean Code lesson. And all I did was ask the reader to figure out how to install the module in an exercise. I wasn't sure they would editable install.

Reply:
i think it's a well placed reminder, maybe an opportunity to give a "further info" lead by being like "unless you have used an [editable install](link to wherever)" but very optional

Reply:
Whenever we can link to the packaging guide or other tutorials, it's great to do so!! it really helps people find our content AND search engines too (altho that is all changing). Cross linking tho is always good!! ✨ 

It's a good idea just to reinforce editable installs here, too! EVERY TIME I / we teach packaging, we get questions about what `pip install` // `pip install -e` // etc is, and pip install (from PyPI), etc. It's an understandable point of confusion. so I'd lean into editable installs, and let's just include a breakout about what the -e is. 

people also get confused about pip install package-name and why they can't pip install their current (local) package name :)